### ISSUE TITLE
Enable SCons shared cache for CIs

### ISSUE THREAD

**Issue by rraallvv** [godotengine/godot#13301]
I was tinkering with AppVeyor to see if enabling the [**SCons shared cache**](http://scons.org/doc/2.0.1/HTML/scons-user/c4213.html) could reduce the build time. This has already been done for Travis with  [**ccache**](https://github.com/ccache/ccache), the problem was that **ccache** does not support MSVC. There is an alternative to **ccache** for MSVC though ([**clcache**](https://github.com/frerich/clcache)), so I decided to give it a try.

**clcache** is a python script so I was just curios because I was thinking that it'll never be as fast as a binary specially optimized that like **ccache**. Anyway, with **clcache** the build time was increased about 40% when building the cache for the first time, and then it decreased about 25% for subsequents builds, both percentages are relative to the build time when the cache is disabled.

Then I decided to try **SCons shared cache**, and boy the freaking thing is fast! Building the cache din't have a noticeable impact, but once the cache is built, the build time is reduced about 75%.

I was still skeptical so I decided to enable **SCons shared cache** for Travis too. Travis already have **ccache** enable, so I wanted to compare the two tools, and **SCons shared cache** seems even faster than **ccache**. Once the cache is built the build time is reduced to a half of what **ccache** takes.

Cheers!

**Comment by rraallvv** [comment]
[PR master](https://github.com/godotengine/godot/pull/13299)
[PR 2.1](https://github.com/godotengine/godot/pull/13300)

**Comment by vnen** [comment]
I believe the Scons cache only make rebuilds faster. Or does AppVeyor keeps the cache between builds?

**Comment by rraallvv** [comment]
@vnen yes Appveyor save/restore the list of directories in the [cache section](https://github.com/godotengine/godot/pull/13299/files#diff-11c909939117928998b102a1fff7d363R14), but that's not enough, also the path has to be specified by [calling CacheDir()](https://github.com/godotengine/godot/pull/13299/files#diff-cc8093822b396591a5530a463c87d048R439) in SConstruct

**Comment by rraallvv** [comment]
I went ahead and made the patches for travis too, just in case. Those disable ccache, but SCons shared cache seem way faster, so it might give room for more builds in the matrix. I didn't touch the matrix though.

[PR master](https://github.com/godotengine/godot/pull/13303)
[PR 2.1](https://github.com/godotengine/godot/pull/13304)

**Comment by akien-mga** [comment]
Fixed with the PRs referenced above. Great work!

**Comment by rraallvv** [comment]
@akien-mga this will take care of pruning the cache to keep it capped to a maximum size specified in the environment variables, cheers!

[PR master](https://github.com/godotengine/godot/pull/13328)
[PR 2.1](https://github.com/godotengine/godot/pull/13329)

**Comment by akien-mga** [comment]
Android fails again with pruning: https://travis-ci.org/godotengine/godot/jobs/307756708

**Comment by rraallvv** [comment]
@akien-mga it seems the problem was the option `progress` passed to SCons in the command line, if set to 'no' the pruning was not being performed. Now the pruning is performed independently of whether the progress indicator is active or not. 

Also the cache size and the umber of files purged from the cache are only shown if the option `verbose` is set to 'yes'.

The cache size and how often it's pruned are controlled by `SCONS_CACHE_LIMIT` in the environs variables and `node_pruning` in **SConstruct**.

The build time would depend mostly on the cache size limit and how often the cache is pruned. `node_pruning = 8` seems ok to me as suggested [here](https://bitbucket.org/scons/scons/wiki/LimitCacheSizeWithProgress), although I don't have enough data to make an educated guess for that value.

The binaries (libraries and executables) are not added to the cache, since those are too big and change too often. Only the intermediate build artifacts are added to the cache.

For the size limit, I think you are right 128 MB is too small, although AppVeyor has a limit of 1024 MB for free accounts and doesn't support separate caches for each branch. So I had to give each branch 512 MB.

Travis allows separate a cache per job and brach so 1024 MB seems ok for Travis, since the big binary files are not added to the cache. I've set the limit for Travis to 1GB, giving a total of 9 GB for all jobs in the master branch and 2.1 branch.

Being able to customize the pruning process with a class seems advantageous too, since we could simply stick to the implementation suggested in the SCons's wiki, or maybe change to some other implementation in the future.

Cheers!
